### PR TITLE
feat(copilot): delegate to the installed Copilot desktop app over its local WS (#86)

### PR DESCRIPTION
**Pull request by robert-crandall**
Part of epic #71. Implements **A (#86)** of the Copilot-app integration reframe. Builds on the docs reframe (#89).

## What

Delegating a task now tries the **installed GitHub Copilot desktop app** over its local WebSocket, and falls back to a cloud `gh agent-task` otherwise. The app's *unofficial* WS protocol is contained behind one adapter (`src/main/agent/copilot-app/`) so an app update can only break one seam; cloud stays the resilient fallback.

The app path ships behind a **feature flag (default OFF)** - with it off, `copilot:delegate` behaves exactly like the old cloud launch. So this PR is a safe merge slice; the delegated-session visibility (link + live status on the todo) lands in the next PR (B / #87).

## The adapter (`src/main/agent/copilot-app/`)

- **discover.ts** - reads `~/.copilot/run/ws.{port,token}` (read-only; the 43-char token is never logged and never put in an error).
- **protocol.ts** - pure builders/parsers for the `create_session` / `send_message` / `session_created` / `session_event` frames.
- **cwd.ts** - trusted local-checkout resolver. Returns a `cwd` only when the path exists, is inside a git worktree, **and a git remote normalizes to the exact `owner/repo`**. That remote check makes the repo-name-under-a-root convention safe against owner collisions.
- **client.ts** - one `ws` connection (no Origin header). **Idempotency boundary:** cloud fallback is only allowed *before* `create_session` is sent; after that, an ambiguous failure never falls back (no double-delegation).
- **delegate.ts** - the WS → cloud strategy with a discriminated `DelegateResult`. A base-branch request routes to cloud (the app WS has no branch concept).
- **store.ts** - a **dedicated `copilot_app_sessions` table** so app sessions can't leak into any existing github-session reader (proven by a reader-isolation test).
- **settings.ts / project-cwd.ts** - the app-delegate flag + the "repos root" convention (default `~/repos`, from the owner's confirmed UX) and a per-project override.

## Wiring

`copilot:delegate` / `copilot:delegate-availability` / `copilot:open-app-session` IPC + a Settings section (flag toggle + repos-root input). `DelegateComposer` now calls `copilot:delegate`.

## Verification

- **45 tests**: protocol, discover, cwd (incl. remote-match edge cases: nested paths, doubled/empty segments, `.git/`), the delegate failure matrix + idempotency boundary, the ws client against a fake server (handshake with no Origin, create/send, pre-create vs post-create classification), and reader isolation (app sessions invisible to `getSessionsForProject` / `getUnassignedSessions` / `getAllStatuses` / `listProjects` / `getProject`).
- **Full suite green** (431 tests), typecheck, lint, and a production build.
- **Real end-to-end**: drove the actual `client.ts` against the running Copilot app with a throwaway prompt + throwaway cwd (create → send → observe → `delete_session`), cleaned up - 0 rows left.
- Reviewed via GPT-5.5 (3 rounds; fixed a remote-normalization false-match, base-branch routing, and the ambiguous-failure UX copy).

## Honest about the seams

The WS is the app's unofficial protocol - it only works when the app is running (and, for the app path, when the repo is checked out under the repos root and validates), and the token rotates per app launch (re-read every connect). Cloud `gh agent-task` is the resilient fallback for everything else.

### CODE REVIEW

File: src/shared/ipc-channels.ts [hunk 1]
```diff
@@ -83,6 +83,60 @@ export interface LaunchTarget {
...
+export interface CopilotAppSession {
+  id: string                     // WS session_id (uuid)
+  projectId: number | null
+  cwd: string                    // the local checkout the app runs in
+  title: string
+  status: CopilotAppSessionStatus
+  repoOwner: string | null
+  repoName: string | null
+  createdAt: string              // ISO 8601
+  updatedAt: string              // ISO 8601
```
[the remark this answers is not in the document]
Comment (robert-crandall) [reply src/shared/ipc-channels.ts]: Fixed in f55560e - these are SQLite datetime('now') values (space-separated UTC), so I corrected the type comments to say SQLite datetime rather than ISO 8601.

File: src/renderer/src/pages/SettingsView.tsx [hunk 2]
```diff
@@ -222,6 +236,49 @@ export function SettingsView({ theme, onClose, onOpenRules }: SettingsViewProps)
...
+                value={reposRoot ?? ''}
+                placeholder="~/repos"
+                autoComplete="off"
+                autoCorrect="off"
+                autoCapitalize="off"
+                spellCheck={false}
+                disabled={reposRoot === null}
+                onChange={(e) => setReposRoot(e.target.value)}
+                onBlur={saveReposRoot}
+              />
```
[the remark this answers is not in the document]
Comment (robert-crandall) [reply src/renderer/src/pages/SettingsView.tsx:274]: Fixed in f55560e - dropped the onBlur handler so the repos-root only saves via the Save button (single mechanism, matching the token field), removing the duplicate IPC call.

File: src/shared/ipc-channels.ts [hunk 3]
```diff
@@ -840,6 +894,61 @@ export type IpcChannels = {
     result: LaunchTarget[]
   }
 
+  /**
+   * Delegate a task: try the installed Copilot desktop app over its local WS
+   * (when the feature flag is on, the app is running, and a trusted local
+   * checkout resolves for the repo), else fall back to a cloud `gh agent-task`.
+   * Returns a discriminated `DelegateResult`. Rejects with an `Error` whose
+   * `message` is `'GH_NOT_AUTHENTICATED'` or starts with `'DELEGATE_FAILED: '`.
+   */
```
[the remark this answers is not in the document]
Comment (robert-crandall) [reply src/shared/ipc-channels.ts:908]: Fixed in 70b9ea8 - documented that the cloud fallback path can reject with LAUNCH_FAILED (from launchAgentTask) in addition to DELEGATE_FAILED; the renderer already strips both prefixes.

File: src/renderer/src/pages/SettingsView.module.css [hunk 4]
```diff
@@ -72,6 +72,22 @@
...
+  color: var(--text-3);
+}
+
+.checkbox {
+  width: 16px;
+  height: 16px;
+  flex-shrink: 0;
+  accent-color: var(--accent);
+  cursor: pointer;
+}
```
[the remark this answers is not in the document]
Comment (robert-crandall) [reply src/renderer/src/pages/SettingsView.module.css:89]: Fixed in 70b9ea8 - added a .checkbox:disabled rule so the disabled state (during initial load) no longer shows an interactive pointer cursor.

File: src/main/agent/copilot-app/discover.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+ * connect — the token rotates per app launch.
+ */
+export function discoverWsEndpoint(runDir: string = copilotRunDir()): WsEndpoint | null {
+  const portRaw = firstLine(join(runDir, 'ws.port'))
+  const token = firstLine(join(runDir, 'ws.token'))
+  if (portRaw === null || token === null) return null
+
+  const port = Number(portRaw)
+  if (!Number.isInteger(port) || port <= 0 || port > 65535) return null
+  if (token.length === 0) return null
```
[the remark this answers is not in the document]
Comment (robert-crandall) [reply src/main/agent/copilot-app/discover.ts]: Fixed in f72de03 - added a digits-only guard and base-10 parse so non-decimal port strings are rejected.

File: src/shared/ipc-channels.ts [hunk 6]
```diff
@@ -83,6 +83,60 @@ export interface LaunchTarget {
...
+export type DelegateResult =
+  | { kind: 'app'; session: CopilotAppSession }
+  | { kind: 'app-send-failed'; session: CopilotAppSession }
+  | { kind: 'cloud'; session: CopilotSession }
+
+/** Why the desktop-app path wasn't taken for a delegate (diagnostic, non-fatal). */
+export type AppDelegateSkipReason =
+  | 'flag_disabled'   // the app-delegate feature flag is off
+  | 'app_not_running' // no WS discovery / connection refused
+  | 'no_local_cwd'    // no trusted local checkout resolved for the repo
```
[the remark this answers is not in the document]
Comment (robert-crandall) [reply src/shared/ipc-channels.ts:137]: Fixed in f72de03 - clarified that app_not_running means the discovery files are absent; no connection is attempted.

File: src/shared/ipc-channels.ts [hunk 7]
```diff
@@ -840,6 +894,63 @@ export type IpcChannels = {
...
+  'copilot:delegate': {
+    args: [payload: DelegatePayload]
+    result: DelegateResult
+  }
+
+  /**
+   * Whether the desktop-app delegate path is currently available for a repo
+   * (flag on + app running + a trusted local checkout resolves). Lets the UI
+   * hint where a delegate will land. Never throws; returns a skip reason.
+   */
```
[the remark this answers is not in the document]
Comment (robert-crandall) [reply src/shared/ipc-channels.ts:920]: Fixed in f72de03 - reworded the docstring to say this is a best-effort hint (no handshake), so a real delegate can still fall back to cloud on e.g. a stale token.

File: src/main/agent/copilot-app/store.ts [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    projectId: row.project_id,
+    cwd: row.cwd,
+    title: row.title,
+    status: row.status as CopilotAppSessionStatus,
+    repoOwner: row.repo_owner,
+    repoName: row.repo_name,
+    createdAt: row.created_at,
+    updatedAt: row.updated_at,
+  }
+}
```
[the remark this answers is not in the document]
Comment (robert-crandall) [reply src/main/agent/copilot-app/store.ts:38]: Fixed in f72de03 - typed AppSessionRow.status as CopilotAppSessionStatus and dropped the cast.

File: src/renderer/src/pages/FocusPage.tsx [hunk 9]
```diff
@@ -164,12 +164,24 @@ export function FocusPage(props: FocusPageProps): JSX.Element {
...
+            } else {
+              const message =
+                result.kind === 'app-send-failed'
+                  ? 'Opened a Copilot session, but I couldn’t hand off the task — open it to retry.'
+                  : 'Copilot is on it in the desktop app.'
+              props.showUndo(
+                message,
+                () => { fire(window.electron.ipc.invoke('copilot:open-app-session', result.session.id)) },
+                'Open'
+              )
```
[the remark this answers is not in the document]
Comment (robert-crandall) [reply src/renderer/src/pages/FocusPage.tsx:183]: Fixed in e31a318 - passed a 'copilot:open-app-session' context so the log is traceable.

File: src/main/agent/copilot-app/delegate.ts [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+function validatePayload(payload: DelegatePayload): { prompt: string; owner: string; name: string } {
+  const prompt = payload.prompt.trim()
+  if (prompt.length === 0) throw new Error('DELEGATE_FAILED: a prompt is required')
+  const owner = payload.repoOwner.trim()
+  const name = payload.repoName.trim()
+  if (owner.length === 0 || name.length === 0) {
+    throw new Error('DELEGATE_FAILED: a target repository is required')
+  }
+  return { prompt, owner, name }
+}
```
[the remark this answers is not in the document]
Comment (robert-crandall) [reply src/main/agent/copilot-app/delegate.ts]: Fixed in e31a318 - validatePayload now returns only { prompt }; owner/name are still validated for presence.

File: src/main/agent/copilot-app/cwd.ts [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+  }
+}
+
+const defaultGitInspector: GitInspector = (dirPath) => {
+  const run = (args: string[]): string | null => {
+    try {
+      return execFileSync('git', ['-C', dirPath, ...args], {
+        encoding: 'utf8',
+        stdio: ['ignore', 'pipe', 'ignore'],
+        timeout: 5000,
```
[the remark this answers is not in the document]
Comment (robert-crandall) [reply src/main/agent/copilot-app/cwd.ts]: Fixed in 1a52c5c - the git inspection now uses promisified execFile (matching github-source.ts), and resolveLocalCwd / the delegate resolveCwd dep / appDelegateAvailability are async, so a hung git can't block the main-thread event loop.

File: src/renderer/src/pages/SettingsView.tsx [hunk 12]
```diff
@@ -87,6 +93,14 @@ export function SettingsView({ theme, onClose, onOpenRules }: SettingsViewProps)
...
     fire(window.electron.ipc.invoke('settings:set-max-sync-days', value), 'settings:set-max-sync-days')
   }
+  const changeAppDelegate = (value: boolean): void => {
+    setAppDelegate(value)
+    fire(window.electron.ipc.invoke('settings:set-app-delegate-enabled', value), 'settings:set-app-delegate-enabled')
+  }
+  const saveReposRoot = (): void => {
+    if (reposRoot === null) return
+    fire(window.electron.ipc.invoke('settings:set-repos-root', reposRoot), 'settings:set-repos-root')
+  }
```
[the remark this answers is not in the document]
Comment (robert-crandall) [reply src/renderer/src/pages/SettingsView.tsx:108]: Fixed in 6a297df - the renderer now normalizes a blank repos-root to DEFAULT_REPOS_ROOT (hoisted into shared/ipc-channels so main and renderer agree) and writes it back to the input, so the stored value shows immediately.
